src/librustc_metadata/lib.rs

```diff
@@ -62,3 +62,5 @@ pub mod index;
 pub mod loader;
 pub mod macro_import;
 pub mod tls_context;
+
+__build_diagnostic_array! { librustc_metadata, DIAGNOSTICS }
```

Comment:
I mentioned this in a comment in the #32777  thread - 

The diagnostics in `librustc_metadata` were not registered in the first function you mentioned, so `rustc --explain` said there was no extended information for any of the errors in `librustc_metadata`.

The line in `metadata/lib.rs` aggregates the diagnostic info into an array, and the one in `driver/lib.rs` registers the information in that array.

With these changes, `rustc --explain` works for all of the error codes in `librustc_metadata`